src/Transports.AspNetCore/WebSockets/BaseSubscriptionServer.cs

```diff
@@ -0,0 +1,488 @@
+                await OnSendKeepAliveAsync();
+            }
+        }
+
+        /*
+         * This 'smart' keep-alive logic doesn't work with subscription-transport-ws because the client will
+         * terminate a connection 20 seconds after the last keep-alive was received (or it will never
+         * terminate if no keep-alive was ever received).
+         * 
+         * See: https://github.com/graphql/graphql-playground/issues/1247
+         */
+        async Task StartSmartKeepAliveLoopAsync()
+        {
+            var lastKeepAliveSent = Connection.LastMessageSentAt;
+            while (true)
+            {
+                var lastSent = Connection.LastMessageSentAt;
+                var lastComm = lastKeepAliveSent > lastSent ? lastKeepAliveSent : lastSent;
+                var now = DateTime.UtcNow;
+                var timePassed = now.Subtract(lastComm);
+                if (timePassed >= keepAliveTimeout)
+                {
+                    await OnSendKeepAliveAsync();
+                    lastKeepAliveSent = now;
+                    await Task.Delay(keepAliveTimeout, CancellationToken);
+                }
+                else
+                {
+                    var timeToWait = keepAliveTimeout - timePassed;
+                    await Task.Delay(timeToWait, CancellationToken);
+                }
+            }
+        }
+    }
+
+    /// <summary>
+    /// Executes when a keep-alive message needs to be sent.
+    /// This method may execute concurrently with other code in this class.
+    /// </summary>
+    protected abstract Task OnSendKeepAliveAsync();
+
+    /// <summary>
+    /// Executes when a connection request needs to be acknowledged.
+    /// </summary>
+    protected abstract Task OnConnectionAcknowledgeAsync(OperationMessage message);
+
+    /// <summary>
+    /// Executes when a new subscription request has occurred.
+    /// Optionally disconnects any existing subscription associated with the same id.
+    /// </summary>
+    protected virtual async Task SubscribeAsync(OperationMessage message, bool overwrite)
+    {
+        if (string.IsNullOrEmpty(message.Id))
+        {
+            await ErrorIdCannotBeBlankAsync(message);
+            return;
+        }
+        var messageId = message.Id!;
+
+        var dummyDisposer = new DummyDisposer();
+
+        try
+        {
+            if (overwrite)
+            {
+                Subscriptions[messageId] = dummyDisposer;
+            }
+            else
+            {
+                if (!Subscriptions.TryAdd(messageId, dummyDisposer))
+                {
+                    await ErrorIdAlreadyExistsAsync(message);
+                    return;
+                }
+            }
+
+            var result = await ExecuteRequestAsync(message);
+            CancellationToken.ThrowIfCancellationRequested();
+            if (!Subscriptions.Contains(messageId, dummyDisposer))
+                return;
```

Comment:
concurrency conditions?

Reply:
And what does this condition mean? I see 3 places of use.

Reply:
`return` would be hit if the client had both requested cancellation of the subscription, and then started a new one with the same id.  According to spec, once the use requests cancellation of the subscription, no more responses must be given -- including single responses or error responses.  So because I have to await the DocumentExecuter to set up the IObservable to begin with, I register a new instance of a temporary object (dummyDisposer) temporarily, and use these special checks to validate that it is still the active subscription for that id.

Reply:
(Sorry it seems so complicated, but I want to implement it correctly.)

Reply:
This definitely needs to be noted in the comments.

Reply:
Ok, I'll work on improving comments within this code block.

Reply:
#818